src/plugins/aws/idc/index.ts

```diff
@@ -0,0 +1,194 @@
+      const init = {
+        method: "POST",
+        body: JSON.stringify({
+          clientName: "p0Cli",
+          clientType: "public",
+          grantTypes: [DEVICE_GRANT_TYPE],
+          scopes: AWS_SSO_SCOPES,
+        }),
+      };
+      const response = await fetch(
+        `https://oidc.${region}.amazonaws.com/client/register`,
+        init
+      );
+      return await response.json();
+    },
+    { duration: AWS_CLIENT_TOKEN_EXPIRY }, // the token has lifetime of 90 days
+    (data) => data.clientSecretExpiresAt < Date.now()
+  );
+
+const awsIdcHelpers = (
+  clientCredentials: { clientId: string; clientSecret: string },
+  idc: { id: string; region: string }
+) => {
+  const { clientId, clientSecret } = clientCredentials;
+  const { id, region } = idc;
+
+  const buildOidcAuthorizeRequest = () => {
+    return {
+      init: {
+        method: "POST",
+        body: JSON.stringify({
+          clientId,
+          clientSecret,
+          startUrl: `https://${id}.awsapps.com/start`,
+        }),
+      },
+      url: `https://oidc.${region}.amazonaws.com/device_authorization`,
+    };
+  };
+  const buildIdcTokenRequest = (authorizeResponse: AWSAuthorizeResponse) => {
+    return {
+      url: `https://oidc.${region}.amazonaws.com/token`,
+      init: {
+        method: "POST",
+        body: JSON.stringify({
+          clientId,
+          clientSecret,
+          deviceCode: authorizeResponse.deviceCode,
+          grantType: DEVICE_GRANT_TYPE,
+        }),
+      },
+    };
+  };
+  /**
+   * Exchanges the oidc token for AWS credentials for a given account and permission set
+   * @param oidcResponse oidc token response fot he oidc /token endpoint
+   * @param request accountId and permissionSet to exchange for AWS credentials
+   * @returns
+   */
+  const exchangeForAwsCredentials = async (
+    oidcResponse: AWSTokenResponse,
+    request: { accountId?: string; permissionSet: string }
+  ) => {
+    // There is a delay in between aws issuing sso token and it to be exchanged for AWS credentials
+    // encountered unauthorized error when exchanging token immediately
+    return await retryWithSleep(
+      async () => {
+        const init = {
+          method: "GET",
+          headers: {
+            "x-amz-sso_bearer_token": oidcResponse.accessToken,
+          },
+        };
+        const { accountId, permissionSet } = request;
+        if (accountId === undefined)
+          throw new Error("Could not find account Id for AWS SSO");
+        const params = new URLSearchParams();
+        params.append("account_id", accountId);
+        params.append("role_name", permissionSet);
+        const response = await fetch(
+          `https://portal.sso.${region}.amazonaws.com/federation/credentials?${params.toString()}`,
+          init
+        );
+        if (!response.ok)
+          throw new Error(
+            `Failed to fetch AWS credentials: ${response.statusText}`
```

Comment:
`statusText` is probably incomplete (it'll just say something like "BAD_REQUEST"). Instead, you probably want
```suggestion
            `Failed to fetch AWS credentials: ${response.statusText}: ${await response.text()}`
```